File: packages/shared/src/engine/rolePolicy.ts
/**
 * Phase 13 Session 3 — the canonical fixed-role authorization *policy*:
 * "may this role ever perform this category of action?" This module
 * answers only that question (architecture doc §14's item #2). It is
 * deliberately silent on the other three layers that gate a real action:
 * visibility (#1, N/A per Phase 13 — every authenticated user sees every
 * project), workflow state (#3, `status.ts`'s `ALLOWED_NEXT`/
 * `canTransitionTo` and the laboratory/stability lifecycle modules), and
 * required-approval-gate state (#4, `ApprovalRecord`/the future workflow
 * engine). `can(role, area, capability) === true` never by itself means an
 * action is currently executable — Session 4 is where role capability,
 * workflow state, and gate state are combined at an actual privileged-
 * action boundary. See docs/PHASE13_IDENTITY_SECURITY_ARCHITECTURE.md §7.
 *
 * Default-deny by construction: `can()` looks a role/area/capability triple
 * up in a fixed, fully-enumerated table and returns `false` for anything
 * not explicitly present — an unknown area, an unsupported capability, or
 * a role string that isn't one of the 12 fixed roles all fall through the
 * same `undefined` lookup path to `false`. There is no "if not explicitly
 * denied, allow" branch anywhere in this file.
 *
 * `Role` is a re-export of `status.ts`'s `ApprovalRole` — deliberately not
 * a second, independently-declared string union. `identity.rs` (Rust) is
 * the storage-layer mirror of the same 12 strings; the parity mechanism
 * (`rolePolicy.roleVocabularyParity.test.ts` + `identity.rs`'s own
 * `role_vocabulary_matches_the_shared_json_fixture` test) is a shared JSON
 * fixture (`roleVocabulary.json`) both languages check *themselves*
 * against, not a third hand-copied list either language trusts the other
 * to keep in sync.
 */
import { APPROVAL_AUTHORITY, APPROVAL_ROLES, type ApprovalRole } from "../schemas/status";
import { AUTHORIZED_REGULATORY_ROLES } from "./regulatoryAuthorization";
import { LABORATORY_METHOD_MANAGER_ROLES } from "./laboratoryStandards";

/** The canonical role vocabulary — imported, not re-declared. */
export type Role = ApprovalRole;
export const ROLES = APPROVAL_ROLES;

/**
 * The real FormuLab workspaces/domains, one per row of the architecture
 * doc's §6 matrix (`pilot_approved`/`production_approved` kept as two
 * distinct areas, `approvalPilot`/`approvalProduction`, exactly mirroring
 * the two separate table rows — they have different worker/approver role
 * sets and must not be merged into one generic "approval" bucket).
 * Deliberately does **not** add a separate "Dossier / Evidence" area: the
 * approved §6 matrix models dossier/claims/labels work as part of the
 * `regulatory` row (see its own header, "Regulatory (dossier, claims/
 * labels)"), and splitting it here would be redesigning the approved
 * matrix rather than transcribing it — flagged, not silently changed
 * (architecture doc §6's own "not yet domain-expert-reviewed" caveat
 * still applies to this grouping).
 */
export const POLICY_AREAS = [
  "home",
  "projects",
  "formulation",
  "laboratory",
  "stability",
  "optimization",
  "rawMaterials",
  "supplierDocuments",
  "regulatory",
  "approvalPilot",
  "approvalProduction",
  "productionEngineering",
  "production",
  "documentControl",
  "reports",
  "dataExchange",
  "administrationUsers",
  "administrationSecurity",
  "administrationSettings",
  "systemAdministration",
] as const;
export type PolicyArea = (typeof POLICY_AREAS)[number];

/** Human-readable labels for the areas above — for a future read-only
 *  "role capabilities" view (Administration → Users, Session 5), never a
 *  second permission source: always generated from `capabilitiesFor`,
 *  never hand-maintained in parallel. */
export const POLICY_AREA_LABELS: Record<PolicyArea, string> = {
  home: "Home",
  projects: "Projects",
  formulation: "Formulation",
  laboratory: "Laboratory (trials, methods, corrective actions)",
  stability: "Stability",
  optimization: "Optimization",
  rawMaterials: "Raw material records (SDS/TDS/COA/specs)",
  supplierDocuments: "Supplier/procurement documentation",
  regulatory: "Regulatory (dossier, claims/labels)",
  approvalPilot: "Approval — pilot_approved",
  approvalProduction: "Approval — production_approved",
  productionEngineering: "Production engineering (scale-up, process prep)",
  production: "Production (batch/process records)",
  documentControl: "Document control (publication, revision, export)",
  reports: "Reports",
  dataExchange: "Data Exchange (import/commit masterdata)",
  administrationUsers: "Administration → Users",
  administrationSecurity: "Administration → Security history",
  administrationSettings: "Administration → App settings",
  systemAdministration:
    "System administration (backup, restore, schema migration, data-location changes)",
};

/**
 * A compact, fixed capability vocabulary (architecture doc §6's legend).
 * `release` is not its own matrix cell — it is the `production` area's
 * `approve` grant under its workflow name (`production_manager`'s
 * batch-release sign-off, §15.4 gate #4); wherever `approve` is granted on
 * `production`, `release` is granted alongside it, from the same cell, not
 * a second hand-typed list (see the `production` row below).
 */
export const CAPABILITIES = [
  "view",
  "create",
  "edit",
  "delete",
  "submit",
  "approve",
  "reject",
  "verify",
  "supersede",
  "release",
  "export",
  "administer",
] as const;
export type Capability = (typeof CAPABILITIES)[number];

type CapabilityGrid = Record<PolicyArea, Record<Role, readonly Capability[]>>;

/** Zips a 12-cell row (in `ROLES`'/`APPROVAL_ROLES`' exact column order —
 *  researcher, research_manager, quality, quality_manager, regulatory,
 *  raw_material, procurement, production_engineering, production,
 *  production_manager, document_control, administrator) into a
 *  `Record<Role, Capability[]>`, so each matrix row below reads as one
 *  line per role in the architecture doc's own column order instead of a
 *  hand-built object literal that could silently transpose two roles. */
function row(cells: readonly (readonly Capability[])[]): Record<Role, readonly Capability[]> {
  if (cells.length !== ROLES.length) {
    throw new Error(`rolePolicy row() expected ${ROLES.length} cells, got ${cells.length}`);
  }
  const out = {} as Record<Role, readonly Capability[]>;
  ROLES.forEach((role, i) => {
    out[role] = cells[i];
  });
  return out;
}

const V: Capability[] = ["view"];
const NONE: Capability[] = [];

/**
 * The full role/area/capability matrix — a direct, cell-by-cell
 * transcription of architecture doc §6's table (Session 1's first full
 * draft, "not yet domain-expert-reviewed" per that doc's own flag,
 * carried forward unchanged here) with two explicit, documented additions
 * beyond the literal §6 cells, both directly instructed by this Session 3
 * brief and by the later, user-approved §15.4/§25 gate-authority decision
 * §6's table itself predates:
 *
 * 1. `production_manager` gains `verify` on `rawMaterials` and
 *    `supplierDocuments` — §6's literal cells for `production_manager` on
 *    those two rows are both "—" (no access at all), which contradicts
 *    §15.4's later, explicit, user-approved decision that
 *    `production_manager` is the sole approval authority for the
 *    raw-material-verification and supplier-document-verification gates.
 *    §15.4 is chronologically later and explicitly user-approved, so it
 *    wins; §6's stale cells are the discrepancy, not this addition — see
 *    docs/PHASE13_IDENTITY_SECURITY_ARCHITECTURE.md §6.3 for the recorded
 *    discrepancy. This grants only `verify` (the gate-approval action),
 *    not `view`/`create`/`edit` — `production_manager` still cannot see or
 *    modify the underlying raw-material/supplier records themselves,
 *    exactly as §6 restricts, only the verification gate decision.
 * 2. `quality` and `administrator` gain `verify` on `regulatory` in addition
 *    to §6's literal "V"-only cells for those two roles —
 *    `regulatoryAuthorization.ts`'s `AUTHORIZED_REGULATORY_ROLES`
 *    (imported below, not re-typed) already, deliberately, grants all
 *    three of `regulatory`/`quality`/`administrator` regulatory-evidence-
 *    confirmation authority today (architecture doc §8: "unchanged...
 *    because that's what the existing, working code already granted").
 *    §6's matrix simply never had this fully reflected in its own cells
 *    (only `regulatory`'s own cell already listed `Vf`); this fixes that
 *    omission against the real, already-enforced authority, it does not
 *    add new authority the app doesn't already have.
 *
 * No other cell deviates from §6's literal table. Everywhere §6 shows "—"
 * (and neither addition above applies), the role has zero capabilities on
 * that area — `can()` returns `false`, not "if not denied, allow".
 */
const MATRIX: CapabilityGrid = {
  home: row([V, V, V, V, V, V, V, V, V, V, V, V]),

  projects: row([
    ["view", "create"],
    ["view", "create", "edit"],
    V,
    V,
    V,
    V,
    V,
    V,
    V,
    V,
    V,
    ["view", "create", "edit", "delete"],
  ]),

  formulation: row([
    ["view", "create", "edit", "submit"],
    ["view", "approve", "reject"],
    V,
    V,
    V,
    V,
    NONE,
    V,
    V,
    V,
    V,
    V,
  ]),

  // `supersede` is added below via a derivation loop from
  // `LABORATORY_METHOD_MANAGER_ROLES` (imported below), not hand-typed
  // into these cells — see the loop after this literal.
  laboratory: row([
    ["view", "create", "edit", "submit"],
    ["view", "approve", "reject"],
    ["view", "create"],
    ["view", "approve", "reject"],
    V,
    V,
    NONE,
    V,
    V,
    V,
    V,
    V,
  ]),

  stability: row([
    ["view", "create", "edit", "submit"],
    ["view", "approve", "reject"],
    ["view", "create"],
    ["view", "approve", "reject"],
    V,
    NONE,
    NONE,
    V,
    NONE,
    V,
    V,
    V,
  ]),

  optimization: row([
    ["view", "create", "edit"],
    V,
    V,
    V,
    V,
    NONE,
    NONE,
    V,
    NONE,
    NONE,
    NONE,
    V,
  ]),

  // Addition #1 (see module doc comment): production_manager -> ["view", "verify"] instead of §6's literal [].
  rawMaterials: row([
    V,
    V,
    ["view", "verify"],
    ["view", "approve"],
    V,
    ["view", "create", "edit"],
    ["view", "create"],
    V,
    NONE,
    ["view", "verify"],
    V,
    V,
  ]),

  // Addition #1 (see module doc comment): production_manager -> ["view", "verify"] instead of §6's literal [].
  supplierDocuments: row([
    NONE,
    NONE,
    V,
    V,
    V,
    V,
    ["view", "create", "edit"],
    NONE,
    NONE,
    ["view", "verify"],
    V,
    V,
  ]),

  // Addition #2 (see module doc comment): `verify` is added below via a
  // derivation loop from `AUTHORIZED_REGULATORY_ROLES` (regulatory,
  // quality, administrator), not hand-typed into these cells.
  regulatory: row([
    V,
    V,
    V,
    V,
    ["view", "create", "edit", "approve", "reject"],
    V,
    V,
    V,
    NONE,
    NONE,
    V,
    V,
  ]),

  // Derived from APPROVAL_AUTHORITY.pilot_approved below, not hand-typed
  // for the approve/reject columns — see the Object.assign block after
  // this literal. "submit" (the preparer/employee tier) has no equivalent
  // in APPROVAL_AUTHORITY (a target-status-keyed approver list only), so
  // it stays a direct §6 transcription.
  approvalPilot: row([
    ["submit"],
    NONE,
    ["submit"],
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
  ]),

  approvalProduction: row([
    NONE,
    NONE,
    ["submit"],
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    ["submit"],
    NONE,
    NONE,
    NONE,
  ]),

  productionEngineering: row([
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    ["view", "create", "edit"],
    V,
    ["view", "approve", "reject"],
    V,
    V,
  ]),

  // `release` rides along with `approve` here — see CAPABILITIES' doc
  // comment: production_manager's batch-release sign-off (§15.4 gate #4)
  // is this cell's `approve`, not a separate grant.
  production: row([
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    V,
    ["view", "create", "edit"],
    ["view", "approve", "reject", "release"],
    V,
    V,
  ]),

  documentControl: row([
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    ["view", "create", "edit", "export", "administer"],
    V,
  ]),

  reports: row([
    ["view", "export"],
    ["view", "export"],
    ["view", "export"],
    ["view", "export"],
    ["view", "export"],
    ["view", "export"],
    ["view", "export"],
    ["view", "export"],
    ["view", "export"],
    ["view", "export"],
    ["view", "export"],
    ["view", "export"],
  ]),

  dataExchange: row([
    ["view", "create"],
    ["view", "create"],
    ["view", "create"],
    ["view", "create"],
    ["view", "create"],
    ["view", "create"],
    ["view", "create"],
    NONE,
    NONE,
    NONE,
    NONE,
    ["view", "create", "administer"],
  ]),

  administrationUsers: row([
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    ["view", "create", "edit", "administer"],
  ]),

  administrationSecurity: row([
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    ["view", "administer"],
  ]),

  administrationSettings: row([
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    ["view", "edit", "administer"],
  ]),

  // Phase 13 Session 4 addition (architecture doc §9.2/§9.3): §6's original
  // table has no System-Administration row at all — Session 3's privileged-
  // command inventory found backup/restore/migration/data-location commands
  // with nothing to enforce against. Per explicit Session 4 instruction,
  // "system-level destructive/configuration mutations should be
  // Administrator-only unless the current approved architecture proves
  // otherwise" — nothing in the approved architecture grants any other role
  // this authority, so `administrator` is the only role with any capability
  // here, exactly like `administrationUsers`/`administrationSecurity` above.
  systemAdministration: row([
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    NONE,
    ["view", "administer"],
  ]),
};

// Every grant below is derived from an existing, independent authority
// source (`LABORATORY_METHOD_MANAGER_ROLES`, `AUTHORIZED_REGULATORY_ROLES`,
// `APPROVAL_AUTHORITY`) instead of re-typing its role list — this is the
// mechanism that makes "approval parity" (architecture doc §19) structural
// rather than a hoped-for convention: change one of those source lists and
// this grid changes with it in the same build, nothing to remember to keep
// in sync by hand.
for (const role of LABORATORY_METHOD_MANAGER_ROLES) {
  MATRIX.laboratory[role] = [...MATRIX.laboratory[role], "supersede"];
}
for (const role of AUTHORIZED_REGULATORY_ROLES) {
  if (!MATRIX.regulatory[role].includes("verify")) {
    MATRIX.regulatory[role] = [...MATRIX.regulatory[role], "verify"];
  }
}
for (const role of APPROVAL_AUTHORITY.pilot_approved) {
  MATRIX.approvalPilot[role] = [...MATRIX.approvalPilot[role], "approve", "reject"];
}
for (const role of APPROVAL_AUTHORITY.production_approved) {
  MATRIX.approvalProduction[role] = [...MATRIX.approvalProduction[role], "approve", "reject"];
}

/**
 * May `role` ever perform `capability` in `area`? Default-deny: an
 * unrecognized area, capability, or role string all fall through the same
 * `undefined` lookup to `false` — there is no "allow unless denied" path.
 *
 * Deliberately typed with plain `string` parameters, not the narrower
 * `Role`/`PolicyArea`/`Capability` unions: real callers usually hold a
 * `SafeUser.role` (`auth.ts`, wire-typed `string` since it crosses the
 * Tauri IPC boundary) or a value from user input/tests, and forcing a cast
 * at every call site would be friction for no safety benefit — the
 * default-deny behavior below is exactly as strict either way. Use the
 * typed `POLICY_AREAS`/`CAPABILITIES`/`ROLES` arrays (or the introspection
 * helpers below, which *are* narrowly typed) when you want compile-time
 * checking of a literal area/capability/role.
 */
export function can(role: string, area: string, capability: string): boolean {
  const areaGrants = (MATRIX as unknown as Record<string, Record<string, readonly string[]> | undefined>)[area];
  if (!areaGrants) return false;
  const roleGrants = areaGrants[role];
  if (!roleGrants) return false;
  return roleGrants.includes(capability);
}

/** Every capability `role` has in `area` — always a real (possibly empty)
 *  array, never `undefined`, since `MATRIX` is exhaustive over every
 *  `PolicyArea` x `Role` pair by construction (`row()` throws at module
 *  load if a row is missing a cell). */
export function capabilitiesFor(role: Role, area: PolicyArea): readonly Capability[] {
  return MATRIX[area][role];
}

/** Every area where `role` has at least one capability — the input a
 *  future read-only "what can this role do" view (Administration → Users,
 *  Session 5) needs, generated from the same canonical policy, never a
 *  second display-only matrix. */
export function areasFor(role: Role): readonly PolicyArea[] {
  return POLICY_AREAS.filter((area) => MATRIX[area][role].length > 0);
}

/** Every role that has `capability` in `area` — e.g. "who can approve
 *  production_approved" for an audit/introspection view. */
export function rolesWithCapability(area: PolicyArea, capability: Capability): readonly Role[] {
  return ROLES.filter((role) => MATRIX[area][role].includes(capability));
}

/**
 * Phase 13 Session 4 — the full, fully-resolved (post-derivation-loop)
 * matrix as a plain, JSON-serializable object: `{ [area]: { [role]:
 * capability[] } }`. This is the one function that turns `MATRIX` — an
 * internal, TypeScript-typed data structure — into the shared contract
 * `scripts/generate-role-policy-matrix.ts` serializes to
 * `rolePolicyMatrix.generated.json`, which `role_policy.rs` (Rust) then
 * reads via `include_str!` to implement its own `can()`. Rust holds no
 * hand-typed permission matrix of its own; this snapshot, generated from
 * this file, is Rust's only source for one. See architecture doc §7/§9.3.
 */
export function fullMatrixSnapshot(): Record<PolicyArea, Record<Role, readonly Capability[]>> {
  const out = {} as Record<PolicyArea, Record<Role, readonly Capability[]>>;
  for (const area of POLICY_AREAS) {
    const roleGrants = {} as Record<Role, readonly Capability[]>;
    for (const role of ROLES) {
      roleGrants[role] = MATRIX[area][role];
    }
    out[area] = roleGrants;
  }
  return out;
}

File: apps/desktop/src-tauri/src/role_policy.rs
// `areas()`/`roles()` are public introspection helpers (mirroring
// `rolePolicy.ts`'s own `POLICY_AREAS`/`ROLES` exports) exercised by this
// module's own tests but not yet consumed elsewhere in the crate — same
// module-level `#[allow(dead_code)]` convention `identity.rs` already uses
// for primitives ahead of their first non-test caller.
#![allow(dead_code)]
// Phase 13 Session 4 — the Rust half of the canonical, cross-language
// authorization policy. This module holds no hand-typed permission matrix
// and no hand-typed workflow-transition graph of its own: both are read at
// first use, via `include_str!`, from the exact same JSON fixtures
// `packages/shared/src/engine/rolePolicy.ts`'s
// `scripts/generate-role-policy-matrix.ts` generates from `MATRIX`/
// `ALLOWED_NEXT` — the identical mechanism `identity.rs`'s
// `role_vocabulary_matches_the_shared_json_fixture` test already uses for
// the 12-role vocabulary (Session 3). A change to `rolePolicy.ts`'s MATRIX
// that isn't followed by regenerating the fixture fails
// `rolePolicy.matrixParity.test.ts` on the TypeScript side, not silently
// here — this file has nothing to independently drift.
//
// See docs/PHASE13_IDENTITY_SECURITY_ARCHITECTURE.md §7/§9.3.
use std::collections::HashMap;
use std::sync::OnceLock;

use serde::Deserialize;

const MATRIX_FIXTURE: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../../packages/shared/src/engine/rolePolicyMatrix.generated.json"
));
const TRANSITIONS_FIXTURE: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../../packages/shared/src/engine/formulaStatusTransitions.json"
));
const MASTERDATA_AREAS_FIXTURE: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../../packages/shared/src/engine/masterdataCollectionAreas.generated.json"
));

#[derive(Debug, Deserialize)]
struct MatrixFixture {
    areas: Vec<String>,
    roles: Vec<String>,
    #[allow(dead_code)] // parsed for completeness; can() only needs `matrix`
    capabilities: Vec<String>,
    matrix: HashMap<String, HashMap<String, Vec<String>>>,
}

#[derive(Debug, Deserialize)]
struct TransitionsFixture {
    #[allow(dead_code)]
    statuses: Vec<String>,
    #[serde(rename = "allowedNext")]
    allowed_next: HashMap<String, Vec<String>>,
}

#[derive(Debug, Deserialize)]
struct MasterdataAreasFixture {
    #[allow(dead_code)]
    collections: Vec<String>,
    areas: HashMap<String, String>,
}

fn matrix() -> &'static MatrixFixture {
    static MATRIX: OnceLock<MatrixFixture> = OnceLock::new();
    MATRIX.get_or_init(|| {
        serde_json::from_str(MATRIX_FIXTURE)
            .expect("rolePolicyMatrix.generated.json must be valid JSON matching MatrixFixture")
    })
}

fn transitions() -> &'static TransitionsFixture {
    static TRANSITIONS: OnceLock<TransitionsFixture> = OnceLock::new();
    TRANSITIONS.get_or_init(|| {
        serde_json::from_str(TRANSITIONS_FIXTURE)
            .expect("formulaStatusTransitions.json must be valid JSON matching TransitionsFixture")
    })
}

fn masterdata_areas() -> &'static MasterdataAreasFixture {
    static AREAS: OnceLock<MasterdataAreasFixture> = OnceLock::new();
    AREAS.get_or_init(|| {
        serde_json::from_str(MASTERDATA_AREAS_FIXTURE).expect(
            "masterdataCollectionAreas.generated.json must be valid JSON matching MasterdataAreasFixture",
        )
    })
}

/// Phase 13 Session 4A (architecture doc §9.3.6/§9.4): the masterdata
/// collection -> `PolicyArea` mapping, read from the same shared fixture
/// `masterdataPolicyAreas.ts`'s `MASTERDATA_COLLECTION_POLICY_AREAS`
/// generates — replaces Session 4's hand-typed Rust `match` with a single
/// source both languages read. Default-deny: an unmapped or unrecognized
/// collection name returns `None`, never a fallback area.
pub fn masterdata_area_for(collection: &str) -> Option<&'static str> {
    masterdata_areas().areas.get(collection).map(|s| s.as_str())
}

/// May `role` ever perform `capability` in `area`? Default-deny, exactly
/// mirroring `rolePolicy.ts`'s `can()`: an unrecognized area, an
/// unrecognized role, or a capability absent from that role's grants in
/// that area all fall through to `false` via plain `HashMap` lookups — there
/// is no "allow unless denied" branch anywhere in this function.
pub fn can(role: &str, area: &str, capability: &str) -> bool {
    matrix()
        .matrix
        .get(area)
        .and_then(|roles| roles.get(role))
        .map(|caps| caps.iter().any(|c| c == capability))
        .unwrap_or(false)
}

/// Every fixed policy area name, for validating an area string before using
/// it elsewhere (e.g. the masterdata collection -> area mapping).
pub fn areas() -> &'static [String] {
    &matrix().areas
}

/// Every fixed role name — the Rust-side mirror of `rolePolicy.ts`'s
/// `ROLES`, sourced from the same fixture `identity::Role` is already
/// checked against (Session 3's `roleVocabulary.json` parity mechanism).
pub fn roles() -> &'static [String] {
    &matrix().roles
}

/// May a `FormulaStatus` move from `from` to `to`? Default-deny: an
/// unrecognized `from` or a `to` not present in its allowed-next list both
/// return `false`. Mirrors `status.ts`'s `ALLOWED_NEXT` graph via the shared
/// fixture — no independently-typed copy of the state machine.
pub fn is_valid_transition(from: &str, to: &str) -> bool {
    transitions()
        .allowed_next
        .get(from)
        .map(|next| next.iter().any(|s| s == to))
        .unwrap_or(false)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_deny_for_unknown_area_role_or_capability() {
        assert!(!can("administrator", "not_a_real_area", "view"));
        assert!(!can("not_a_real_role", "formulation", "view"));
        assert!(!can("administrator", "formulation", "not_a_real_capability"));
        assert!(!can("", "", ""));
    }

    #[test]
    fn every_role_has_view_on_home() {
        for role in roles() {
            assert!(can(role, "home", "view"), "{role} should have view on home");
        }
    }

    #[test]
    fn only_administrator_can_manage_users_or_view_security_history() {
        // Phase 13 Session 5 — the structural proof "non-admin cannot
        // manage users" rests on: no other role has ANY capability in
        // either administration area.
        for role in roles() {
            let expected_admin_only = role == "administrator";
            for capability in ["view", "create", "edit", "administer"] {
                assert_eq!(
                    can(role, "administrationUsers", capability),
                    expected_admin_only,
                    "administrationUsers/{capability} must be administrator-only, got role={role}"
                );
            }
            assert_eq!(
                can(role, "administrationSecurity", "view"),
                expected_admin_only,
                "administrationSecurity/view must be administrator-only, got role={role}"
            );
        }
    }

    #[test]
    fn only_administrator_has_any_capability_on_system_administration() {
        for role in roles() {
            let expected_admin_only = role == "administrator";
            assert_eq!(
                can(role, "systemAdministration", "administer"),
                expected_admin_only,
                "systemAdministration/administer must be administrator-only, got role={role}"
            );
        }
    }

    #[test]
    fn administrator_never_holds_create_or_edit_on_any_scientific_content_area() {
        // Session 6 (privilege-escalation / authorization-bypass suite):
        // the direct, positive-denial proof of architecture doc §9's
        // "administrator is view-only on scientific content by explicit
        // design" rule — not merely inferred from the absence of a
        // positive grant elsewhere. Every area below is scientific/
        // business content administrator must never create or edit,
        // regardless of the narrow, explicit gate-decide exceptions
        // (verify/approve/reject) §15.4 grants on four of them.
        for area in [
            "formulation",
            "laboratory",
            "stability",
            "optimization",
            "rawMaterials",
            "supplierDocuments",
            "regulatory",
            "productionEngineering",
            "production",
        ] {
            for capability in ["create", "edit"] {
                assert!(
                    !can("administrator", area, capability),
                    "administrator must never hold {capability} on {area} — view-only by explicit design (§9)"
                );
            }
            assert!(can("administrator", area, "view"), "administrator should still view {area}");
        }
    }

    #[test]
    fn approval_pilot_approve_matches_the_known_manager_tier_plus_administrator() {
        for role in ["research_manager", "quality_manager", "administrator"] {
            assert!(can(role, "approvalPilot", "approve"), "{role} should approve pilot_approved");
        }
        for role in ["researcher", "quality", "regulatory", "raw_material", "procurement", "production_engineering", "production", "production_manager", "document_control"] {
            assert!(!can(role, "approvalPilot", "approve"), "{role} must NOT approve pilot_approved");
        }
    }

    #[test]
    fn approval_production_approve_matches_the_known_authority_set() {
        for role in ["quality_manager", "regulatory", "production_manager", "administrator"] {
            assert!(can(role, "approvalProduction", "approve"), "{role} should approve production_approved");
        }
        for role in ["researcher", "research_manager", "quality", "raw_material", "procurement", "production_engineering", "production", "document_control"] {
            assert!(!can(role, "approvalProduction", "approve"), "{role} must NOT approve production_approved");
        }
    }

    #[test]
    fn worker_tier_cannot_approve_even_though_it_can_submit() {
        // The exact worker/manager separation Session 4's brief requires
        // proof of: a worker role has `submit` but never `approve`/`reject`.
        assert!(can("researcher", "approvalPilot", "submit"));
        assert!(!can("researcher", "approvalPilot", "approve"));
        assert!(!can("researcher", "approvalPilot", "reject"));
        assert!(can("quality", "approvalProduction", "submit"));
        assert!(!can("quality", "approvalProduction", "approve"));
    }

    #[test]
    fn production_manager_verify_on_raw_materials_and_supplier_documents_session1_closure_gates() {
        // architecture doc §15.4 — the discrepancy-resolution addition.
        assert!(can("production_manager", "rawMaterials", "verify"));
        assert!(can("production_manager", "supplierDocuments", "verify"));
    }

    #[test]
    fn quality_does_not_hold_the_raw_material_gate_decide_capability() {
        // Phase 13 closure-session correction (rolePolicy.ts's own
        // "Correction #4" doc comment): production_manager (plus
        // administrator) is this gate's sole decide authority — `quality`'s
        // stale pre-gate `verify` grant on `rawMaterials` quietly acted as
        // a second one and has been removed from the canonical matrix.
        assert!(!can("quality", "rawMaterials", "verify"));
        assert!(can("quality", "rawMaterials", "view"));
    }

    #[test]
    fn regulatory_verify_extends_to_quality_and_administrator() {
        // architecture doc §8 — AUTHORIZED_REGULATORY_ROLES-derived addition.
        for role in ["regulatory", "quality", "administrator"] {
            assert!(can(role, "regulatory", "verify"), "{role} should verify regulatory evidence");
        }
        assert!(!can("researcher", "regulatory", "verify"));
    }

    #[test]
    fn transition_graph_allows_known_valid_moves_and_denies_invalid_ones() {
        assert!(is_valid_transition("pilot_candidate", "pilot_approved"));
        assert!(is_valid_transition("pilot_approved", "production_approved"));
        assert!(is_valid_transition("concept", "rejected"));
        assert!(is_valid_transition("rejected", "concept"));
        // Not adjacent: concept cannot jump straight to pilot_approved.
        assert!(!is_valid_transition("concept", "pilot_approved"));
        // production_approved's only forward move is retired.
        assert!(!is_valid_transition("production_approved", "pilot_candidate"));
        assert!(is_valid_transition("production_approved", "retired"));
        // retired is terminal.
        assert!(!is_valid_transition("retired", "concept"));
        // Unknown statuses default-deny.
        assert!(!is_valid_transition("not_a_real_status", "concept"));
    }

    #[test]
    fn fixture_vocabularies_have_the_expected_shape() {
        assert_eq!(roles().len(), 12);
        assert!(areas().contains(&"systemAdministration".to_string()));
        assert!(areas().contains(&"formulation".to_string()));
    }

    // -------------------------------------------- masterdata_area_for ---

    #[test]
    fn masterdata_area_for_has_all_91_collections_mapped() {
        assert_eq!(masterdata_areas().collections.len(), 91);
        for collection in &masterdata_areas().collections {
            assert!(
                masterdata_area_for(collection).is_some(),
                "{collection} has no policy area mapping"
            );
        }
    }

    #[test]
    fn masterdata_area_for_denies_an_unknown_collection() {
        assert_eq!(masterdata_area_for("not_a_real_collection"), None);
        assert_eq!(masterdata_area_for(""), None);
        assert_eq!(masterdata_area_for("../../etc/passwd"), None);
    }

    #[test]
    fn masterdata_area_for_matches_representative_collections() {
        assert_eq!(masterdata_area_for("materials"), Some("rawMaterials"));
        assert_eq!(masterdata_area_for("stability_studies"), Some("stability"));
        assert_eq!(masterdata_area_for("test_results"), Some("laboratory"));
        assert_eq!(masterdata_area_for("regulatory_rules"), Some("regulatory"));
        assert_eq!(masterdata_area_for("doe_studies"), Some("optimization"));
        assert_eq!(masterdata_area_for("data_exchange_import_jobs"), Some("dataExchange"));
        assert_eq!(masterdata_area_for("generated_document_records"), Some("documentControl"));
    }

    #[test]
    fn every_masterdata_area_is_a_real_policy_area() {
        let known: std::collections::HashSet<&str> = areas().iter().map(|s| s.as_str()).collect();
        for collection in &masterdata_areas().collections {
            let area = masterdata_area_for(collection).unwrap();
            assert!(known.contains(area), "{collection} -> {area} is not a real PolicyArea");
        }
    }
}
